Problem: (CRO-599) No way to get public key for creating new MultiSig address

devashishdxt opened the pull request:
Solution: Exposed `wallet_newPublicKey` method in `client-rpc`

tomtau commented on client-core/src/hd_seed.rs:
```diff
@@ -39,13 +39,14 @@ impl HDSeed {
-            AddressType::Staking => 1,
+            Some(AddressType::Transfer) => 0,
+            Some(AddressType::Staking) => 1,
+            None => 2,
```
I don't have enough context for this, but "multisig" address is a transfer address, so why this separation? Just create a transfer address and get an associated public key?

devashishdxt replied:
Transfer addresses in HD wallet are 1-of-n transfer addresses which are created using `wallet_client.new_transfer_address()`.

tomtau commented:
The issue / problem was "No way to get public key of transfer address for MultiSig address creation",
not "No way to get public key for creating new MultiSig address"...

I assume it could be solved, say, by API that lists public keys or given an address, it'll return the associated public key or something like that -- @calvinaco ?

devashishdxt commented:
> The issue / problem was "No way to get public key of transfer address for MultiSig address creation",
> not "No way to get public key for creating new MultiSig address"...
> 
> I assume it could be solved, say, by API that lists public keys or given an address, it'll return the associated public key or something like that -- @calvinaco ?

@tomtau Some context about this in comments of this: https://mcoproduct.atlassian.net/browse/CRO-601

calvinlauco commented:
> The issue / problem was "No way to get public key of transfer address for MultiSig address creation",
> not "No way to get public key for creating new MultiSig address"...
> 
> I assume it could be solved, say, by API that lists public keys or given an address, it'll return the associated public key or something like that -- @calvinaco ?

That was my original intention. But the core of the question is actually to re-enable the ability to create a MultiSig address for a wallet, which is broken before this fix.

I have discussed with @devashishdxt  and I think this is more trivial way to re-enable this feature.

devashishdxt commented:
> 1. as multisig addresses are transfer addresses, the extra "account" separation is just an accidental complexity - or what does it help with to have a separate index?
> 2. calling the api "new public key" is a bit confusing (it's in the internals, but that could be renamed with it), because all addresses are derived from public keys, so when I come to as a first time, its purpose is unclear -- perhaps something like `new_transfer_address_component`, `new_transfer_address_multi_part`... (as `new_transfer_address` currently assumes 1-of-1) may make it more explicit

1. Removed an extra index.
2. Renamed RPC function to `new_multi_sig_address_public_key`

tomtau commented:
bors r+

calvinlauco commented:
> > The issue / problem was "No way to get public key of transfer address for MultiSig address creation",
> > not "No way to get public key for creating new MultiSig address"...
> > I assume it could be solved, say, by API that lists public keys or given an address, it'll return the associated public key or something like that -- @calvinaco ?
> 
> That was my original intention. But the core of the question is actually to re-enable the ability to create a MultiSig address for a wallet, which is broken/not available before this fix.
> 
> I have discussed with @devashishdxt and I think this is a trivial way to re-enable this feature.

My intention is trying to simply this create MultiSig address process: With a `createPublicKey` call we need only two steps:
1. create public key
2. use it to create MultiSig address
Done

But then I study again how Bitcoin does it and I have the following thoughts:
https://en.bitcoin.it/wiki/Multisignature#Creating_a_Multisignature_Address_with_Bitcoin-Qt
- Creating a transfer address and query public key from it is acceptable
- The transfer address can be kind of act as a key to query back the public key used?

On the other hand, I agree with @tomtau that MultiSig address is a kind of transfer address. So I don't prefer to separate it from the transfer address index in HD wallet.

@devashishdxt any thought?

devashishdxt commented:
> * Creating a transfer address and query public key from it is acceptable

I don't think we should re-use a public key from an already created transfer address. This may be acceptable but should not be recommended.

> * The transfer address can be kind of act as a key to query back the public key used?

You can easily get your public key from existing multi-sig address (1-of-n/m-of-n). See this:
https://github.com/crypto-com/chain/blob/8af48ceac8c580aa91c89bedc906f06fbba225d7/client-core/src/service/root_hash_service.rs#L67
